test/format/verify.c

```diff
@@ -0,0 +1,386 @@
+      base->type == ROW ? (char *)base_key->data : "#", (int)base_value->size,
+      (char *)base_value->data, base_bitv, table->uri, table_bitv,
+      checkpoint ? " (checkpoint " : "", checkpoint ? checkpoint : "", checkpoint ? ")" : "");
+    fprintf(stderr,
+      "mirror: %" PRIu64 "/%" PRIu64 " mismatch: %s: {%.*s}/{%.*s} [%#x], %s: {#}/{%#x} %s%s%s\n",
+      base_keyno, table_keyno, base->uri, base->type == ROW ? (int)base_key->size : 1,
+      base->type == ROW ? (char *)base_key->data : "#", (int)base_value->size,
+      (char *)base_value->data, base_bitv, table->uri, table_bitv,
+      checkpoint ? " (checkpoint " : "", checkpoint ? checkpoint : "", checkpoint ? ")" : "");
+}
+
+/*
+ * table_verify_mirror --
+ *     Verify a mirrored pair.
+ */
+static void
+table_verify_mirror(WT_CONNECTION *conn, TABLE *base, TABLE *table, const char *checkpoint)
+{
+    SAP sap;
+    WT_CURSOR *base_cursor, *table_cursor;
+    WT_ITEM base_key, base_value, table_key, table_value;
+    WT_SESSION *session;
+    uint64_t base_id, base_keyno, last_match, table_id, table_keyno, rows;
+    uint8_t base_bitv, table_bitv;
+    u_int failures, i;
+    int base_ret, table_ret;
+    char buf[256], tagbuf[128];
+
+    base_keyno = table_keyno = 0;             /* -Wconditional-uninitialized */
+    base_bitv = table_bitv = FIX_VALUE_WRONG; /* -Wconditional-uninitialized */
+    base_ret = table_ret = 0;
+    last_match = 0;
+
+    memset(&sap, 0, sizeof(sap));
+    wt_wrap_open_session(conn, &sap, NULL, &session);
+
+    /* Optionally open a checkpoint to verify. */
+    if (checkpoint != NULL)
+        testutil_check(__wt_snprintf(buf, sizeof(buf), "checkpoint=%s", checkpoint));
+
+    /*
+     * If opening a checkpoint, retry if the cursor checkpoint IDs don't match, it just means that a
+     * checkpoint happened between the two open calls.
+     */
+    for (;;) {
+        wt_wrap_open_cursor(session, base->uri, checkpoint == NULL ? NULL : buf, &base_cursor);
+        base_id = base_cursor->checkpoint_id(base_cursor);
+        wt_wrap_open_cursor(session, table->uri, checkpoint == NULL ? NULL : buf, &table_cursor);
+        table_id = table_cursor->checkpoint_id(table_cursor);
+
+        testutil_assert((checkpoint == NULL && base_id == 0 && table_id == 0) ||
+          (checkpoint != NULL && base_id != 0 && table_id != 0));
+
+        if (checkpoint == NULL || base_id == table_id)
+            break;
+        testutil_check(base_cursor->close(base_cursor));
+        testutil_check(table_cursor->close(table_cursor));
+    }
+
+    testutil_check(__wt_snprintf(buf, sizeof(buf),
+      "table %u %s%s"
+      "mirror verify",
+      table->id, checkpoint == NULL ? "" : checkpoint, checkpoint == NULL ? "" : " checkpoint "));
+    trace_msg(session, "%s: start", buf);
+
+    for (failures = 0, rows = 1; rows <= TV(RUNS_ROWS); ++rows) {
+        switch (base->type) {
+        case FIX:
+            testutil_assert(base->type != FIX);
```

Comment:
I don't quite understand this. It seems that `base->type` can't be FLCS, but I thought it was valid to have a base type of FLCS, and it's the mirrored table that can't be FLCS?

Perhaps I'm misunderstanding the roles of `base` and `table` here, I'd assumed `base` is the "original" table and `table` is the mirror.

Reply:
It's the original table that can't be FLCS; FLCS values don't hold enough information to mirror them usefully. (I suppose it could be done anyway, but it wasn't... also, you could still mirror FLCS to FLCS, but that wasn't done either.)

Reply:
Ah, OK. Thanks for clarifying.